tower-http/src/limit.rs

```diff
@@ -0,0 +1,246 @@
+//! {
+//!     hyper::body::to_bytes(req.into_body()).await?;
+//!     Ok(Response::new(Body::empty()))
+//! }
+//!
+//! let mut svc = ServiceBuilder::new()
+//!     // Limit incoming requests to 4096 bytes.
+//!     .layer(LengthLimitedLayer::new(4096))
+//!     .service_fn(handle);
+//!
+//! fn test_svc<S: Service<Request<Body>>>(s: &S) {}
+//! test_svc(&svc);
+//!
+//! // Call the service.
+//! let request = Request::new(Body::empty());
+//!
+//! let response = svc.ready().await?.call(request).await?;
+//!
+//! assert_eq!(response.status(), 200);
+//!
+//! // Call the service with a body that is too large.
+//! let request = Request::new(Body::from(Bytes::from(vec![0u8; 4097])));
+//!
+//! let response = svc.ready().await?.call(request).await?;
+//!
+//! assert_eq!(response.status(), StatusCode::PAYLOAD_TOO_LARGE);
+//!
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use crate::BoxError;
+use bytes::Bytes;
+use http::{HeaderValue, Request, Response, StatusCode};
+use http_body::combinators::UnsyncBoxBody;
+use http_body::{Body, Full, LengthLimitError, Limited};
+use pin_project_lite::pin_project;
+use std::error::Error;
+use std::future::Future;
+use std::pin::Pin;
+use std::{
+    any, fmt,
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies the [`LengthLimit`] middleware that intercepts requests
+/// with body lengths greater than the configured limit and converts them into
+/// `413 Payload Too Large` responses.
+///
+/// See the [module docs](self) for an example.
+pub struct LengthLimitedLayer<B> {
+    limit: usize,
+    _ty: PhantomData<fn() -> B>,
+}
+
+impl<B> LengthLimitedLayer<B> {
+    /// Create a new `LengthLimitedLayer` with the given body length limit.
+    pub fn new(limit: usize) -> Self {
+        Self {
+            limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<B> Clone for LengthLimitedLayer<B> {
+    fn clone(&self) -> Self {
+        Self {
+            limit: self.limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<B> Copy for LengthLimitedLayer<B> {}
+
+impl<B> fmt::Debug for LengthLimitedLayer<B> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LengthLimitedLayer")
+            .field("body", &any::type_name::<B>())
+            .field("limit", &self.limit)
+            .finish()
+    }
+}
+
+impl<B, S> Layer<S> for LengthLimitedLayer<B> {
+    type Service = LengthLimited<S, B>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        LengthLimited {
+            inner,
+            limit: self.limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+/// Middleware that intercepts requests with body lengths greater than the
+/// configured limit and converts them into `413 Payload Too Large` responses.
+///
+/// See the [module docs](self) for an example.
+pub struct LengthLimited<S, B> {
```

Comment:
What do you think about calling it `RequestBodyLimit`? Then its more clear its limiting the request body and not the response body.